docs/DESIGN.md

```diff
@@ -0,0 +1,164 @@
+    * Minor bugs listed as issues
+    * Ensure the integrity of the test suite!
+
+4. Concurrently with (3), build out examples and tutorials in `/examples`
+
+
+### Backlog
+
+ * Other branches: fd/monad-lift and lots-of-optics
+ * When multiplicity is supported, anything that can be made polymorphic should
+   be made polymorphic -- this will make things much more sane.
+ * There are a lot of old branches that need to be reviewed.
+
+## Overall Architecture
+
+The provided data, functions and classes in linear base can be divided into
+three parts.
+
+1. Variants of tools in `base` (or highly-used common haskell tools) in which
+   some function arrows are replaced with linear arrows.
+  * E.g., `Data.Functor.Linear`'s `Functor` has
+  `fmap :: (a #-> b) -> f a #-> f b`.
+2. Re-designed APIs for tools in `base`.
+  * E.g., file IO is done with a linear resource monad, `RIO` in
+    `System.IO.Resource`
+3. New tools you naturally find yourself wanting with linear types.
+  * E.g., an `Unrestricted` data type.
+  * E.g., mutable arrays in `Data.Array.Mutable.Linear`.
+
+### General Implementation Strategy
+
+This is the strategy (that we've followed so far) for developing linear base:
+
+Variants are reasonably simple to implement. You just add linear arrows where
+you would need them if you were trying to use that base facility in a linear
+function. If there are multiple use cases, you might need to think more about
+the design -- as was the case with [functors]. Here's an example:
+
+```haskell
+foldr :: (a #-> b #-> b) -> b #-> [a] #-> b
+foldr f z = \case
+  [] -> z
+  x:xs -> f x (foldr f z xs)
+```
+
+Re-designed APIs are re-designed because the original API had problems that can
+be solved with linear types. This is the case with file IO.  A concrete example
+is provided in [examples/Simple/FileIO.hs].  To implement something in this
+group, do this:
+
+1. carefully enumerate all the unsafe things about the old API
+2. design a safe new linear API and argue (preferably with a proof written down
+   somewhere)
+3. implement your API
+
+The ideas behind new tools vary and each have unique concepts that are not
+addressed by a general strategy. These should be documented in the section
+below.
+
+## How Key Pieces are Implemented
+
+### Num and similar classes
+
+We can use `DerivingVia` and `newtype`s to get instances of linear versions of
+classes. See [`Num`] for details.
+
+### Mutable Data Structures
+
+Mutable data structures release constructors that take continuations.  The
+accessors linearly thread the data structure.  The mutators linearly consume
+and return the data structure.  These are always consumable.
+
+```haskell
+allocate :: Int -> (DataStructure a #-> Unrestricted b) #-> Unrestricted b
+mutate :: Something -> DataStructure a #-> DataStructure a
+access :: Something -> DataStructure a #-> (DataStructure a, AccessInfo)
+instance Consumable (DataStructure a)
+```
+
+## Rules and Advice on Contributing
```

Comment:
There is a CONTRIBUTING.md here: https://github.com/tweag/.github/blob/master/CONTRIBUTING.md

I think it may be picked up automatically by Github when someone writes a PR. I don't know how this `.github` thing works. Ask around. But at the end of the day, for everything that is not specific to linear-base, just point to them rather than repeating.